system/lib/libc/emscripten_mmap.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2022 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+#include <assert.h>
+#include <errno.h>
+#include <malloc.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+
+#include <emscripten/heap.h>
+#include <emscripten/emmalloc.h>
+
+#include "lock.h"
+#include "syscall.h"
+
+struct map {
+  void* addr;
+  long length;
+  int allocated;
+  int fd;
+  int flags;
+  off_t offset;
+  int prot;
+  struct map* next;
+} __attribute__((aligned (1)));
+
+// Linked list of all mapping, guarded by a musl-style lock (LOCK/UNLOCK)
+static volatile int lock[1];
```

Comment:
why `[1]`? that seems unnecessary, but likely is some C thing I don't know...

Reply:
I'm just copying existing musl code:

```
$ git grep "volatile int lock"
system/lib/libc/musl/src/dirent/__dirent.h:     volatile int lock[1];
system/lib/libc/musl/src/exit/at_quick_exit.c:static volatile int lock[1];
system/lib/libc/musl/src/exit/atexit.c:static volatile int lock[1];
system/lib/libc/musl/src/internal/stdio_impl.h: volatile int lock;
system/lib/libc/musl/src/locale/dcngettext.c:static volatile int lock[1];
system/lib/libc/musl/src/misc/syslog.c:static volatile int lock[1];
system/lib/libc/musl/src/network/lookup.h:      volatile int lock[1];
system/lib/libc/musl/src/prng/random.c:static volatile int lock[1];
system/lib/libc/musl/src/thread/pthread_atfork.c:static volatile int lock[1];
system/lib/libc/musl/src/thread/sem_open.c:static volatile int lock[1];
system/lib/libc/musl/src/time/__tz.c:static volatile int lock[1];
system/lib/standalone/standalone.c:static volatile int lock[1];
tests/pthread/test_pthread_gcc_spinlock.cpp:volatile int lock = 0; // spinlock "mutex" variable
````

My guess is that its so that the bare use of the `lock` variable is always a pointer and you don't need to pass `&lock` everywhere?

Reply:
Ah, makes sense...